File: python/src/context_firewall/boundary.py
# SPDX-License-Identifier: BSL-1.1
# Copyright (c) 2026 MuVeraAI Corporation

"""
BoundaryRule definitions for context-firewall.

A BoundaryRule specifies which data types may or may not cross between
two domains. Rules are evaluated by DataInspector before any crossing is permitted.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Protocol

from pydantic import BaseModel, Field

from .types import BoundaryDirection, DataClassification

__all__ = [
    "BoundaryRule",
    "BoundaryRuleConfig",
    "StandardBoundaryRule",
    "create_boundary_rule",
    "HEALTH_TO_WORK_RULE",
    "FINANCIAL_TO_WORK_RULE",
    "HEALTH_TO_PERSONAL_RULE",
    "FINANCIAL_TO_PERSONAL_RULE",
    "PERSONAL_TO_WORK_RULE",
    "DEFAULT_BOUNDARY_RULES",
    "BoundaryRegistry",
    "build_boundary_key",
    "build_boundary_registry",
]

# ---------------------------------------------------------------------------
# BoundaryRule protocol
# ---------------------------------------------------------------------------


class BoundaryRule(Protocol):
    """
    Protocol describing a boundary rule between two domains.

    Implementors can subclass :class:`StandardBoundaryRule` or implement
    this protocol directly for custom evaluation logic.
    """

    @property
    def name(self) -> str:
        """Unique identifier for this rule."""
        ...

    @property
    def from_domain(self) -> str:
        """The originating domain."""
        ...

    @property
    def to_domain(self) -> str:
        """The destination domain."""
        ...

    @property
    def direction(self) -> BoundaryDirection:
        """Direction of the rule: ``one-way`` or ``bidirectional``."""
        ...

    @property
    def allowed_data_types(self) -> list[str]:
        """Data types explicitly allowed to cross. Empty means no allowlist."""
        ...

    @property
    def blocked_data_types(self) -> list[str]:
        """Data types that must never cross."""
        ...

    def evaluate(self, classification: DataClassification) -> bool:
        """
        Evaluate whether a classification is permitted to cross.

        :param classification: The classification of the data being inspected.
        :returns: ``True`` if crossing is allowed, ``False`` if blocked.
        """
        ...


# ---------------------------------------------------------------------------
# BoundaryRuleConfig — serialisable config model
# ---------------------------------------------------------------------------


class BoundaryRuleConfig(BaseModel):
    """Serialisable configuration for a standard boundary rule."""

    name: str = Field(description="Unique rule identifier.")
    from_domain: str = Field(description="Originating domain name.")
    to_domain: str = Field(description="Destination domain name.")
    direction: BoundaryDirection = Field(description="Rule direction.")
    allowed_data_types: list[str] = Field(
        default_factory=list,
        description="Explicitly allowed data types. Empty means no allowlist.",
    )
    blocked_data_types: list[str] = Field(
        default_factory=list,
        description="Data types that are never permitted to cross.",
    )

    model_config = {"frozen": True}


# ---------------------------------------------------------------------------
# StandardBoundaryRule — concrete implementation
# ---------------------------------------------------------------------------


class StandardBoundaryRule:
    """
    Standard implementation of :class:`BoundaryRule`.

    Evaluation logic:

    1. If ``blocked_data_types`` is non-empty, any detected type in that list blocks the crossing.
    2. If ``allowed_data_types`` is non-empty, only types on the list are permitted.
    3. If both lists are empty, the crossing is allowed (open rule).

    Example::

        rule = StandardBoundaryRule(BoundaryRuleConfig(
            name="health->work",
            from_domain="health",
            to_domain="work",
            direction="one-way",
            blocked_data_types=["medical", "prescription", "diagnosis"],
        ))
        rule.evaluate(classification)  # False if medical data detected
    """

    def __init__(self, config: BoundaryRuleConfig) -> None:
        self._config = config

    @property
    def name(self) -> str:
        return self._config.name

    @property
    def from_domain(self) -> str:
        return self._config.from_domain

    @property
    def to_domain(self) -> str:
        return self._config.to_domain

    @property
    def direction(self) -> BoundaryDirection:
        return self._config.direction

    @property
    def allowed_data_types(self) -> list[str]:
        return list(self._config.allowed_data_types)

    @property
    def blocked_data_types(self) -> list[str]:
        return list(self._config.blocked_data_types)

    def evaluate(self, classification: DataClassification) -> bool:
        """Evaluate whether the classification may cross this boundary."""
        blocked = set(self._config.blocked_data_types)
        allowed = set(self._config.allowed_data_types)

        for detected_type in classification.detected_types:
            if detected_type in blocked:
                return False
            if allowed and detected_type not in allowed:
                return False
        return True


# ---------------------------------------------------------------------------
# Factory
# ---------------------------------------------------------------------------


def create_boundary_rule(config: BoundaryRuleConfig) -> StandardBoundaryRule:
    """
    Build a :class:`StandardBoundaryRule` from a :class:`BoundaryRuleConfig`.

    :param config: Serialisable rule configuration.
    :returns: A fully formed rule ready for registration.
    """
    return StandardBoundaryRule(config)


# ---------------------------------------------------------------------------
# Default built-in boundary rules
# ---------------------------------------------------------------------------

HEALTH_TO_WORK_RULE = create_boundary_rule(
    BoundaryRuleConfig(
        name="health->work",
        from_domain="health",
        to_domain="work",
        direction="one-way",
        blocked_data_types=[
            "medical",
            "prescription",
            "diagnosis",
            "mental-health",
            "clinical",
            "laboratory",
        ],
    )
)
"""health -> work: no medical data crosses into professional context."""

FINANCIAL_TO_WORK_RULE = create_boundary_rule(
    BoundaryRuleConfig(
        name="financial->work",
        from_domain="financial",
        to_domain="work",
        direction="one-way",
        blocked_data_types=[
            "banking",
            "credit-card",
            "tax",
            "investment",
            "insurance-financial",
            "salary",
        ],
    )
)
"""financial -> work: no financial account or tax data crosses into work context."""

HEALTH_TO_PERSONAL_RULE = create_boundary_rule(
    BoundaryRuleConfig(
        name="health->personal",
        from_domain="health",
        to_domain="personal",
        direction="one-way",
        blocked_data_types=[
            "medical",
            "prescription",
            "diagnosis",
            "mental-health",
            "clinical",
            "laboratory",
        ],
    )
)
"""health -> personal: medical data may not cross into personal domain by default."""

FINANCIAL_TO_PERSONAL_RULE = create_boundary_rule(
    BoundaryRuleConfig(
        name="financial->personal",
        from_domain="financial",
        to_domain="personal",
        direction="one-way",
        blocked_data_types=[
            "banking",
            "credit-card",
            "tax",
            "investment",
            "salary",
        ],
    )
)
"""financial -> personal: core financial account data may not flow into personal context."""

PERSONAL_TO_WORK_RULE = create_boundary_rule(
    BoundaryRuleConfig(
        name="personal->work",
        from_domain="personal",
        to_domain="work",
        direction="one-way",
        blocked_data_types=[
            "family",
            "relationship",
            "home-address",
            "personal-contact",
        ],
    )
)
"""personal -> work: personal and family data must not enter the work domain."""

DEFAULT_BOUNDARY_RULES: tuple[StandardBoundaryRule, ...] = (
    HEALTH_TO_WORK_RULE,
    FINANCIAL_TO_WORK_RULE,
    HEALTH_TO_PERSONAL_RULE,
    FINANCIAL_TO_PERSONAL_RULE,
    PERSONAL_TO_WORK_RULE,
)
"""The complete set of default built-in boundary rules."""

# ---------------------------------------------------------------------------
# Boundary registry
# ---------------------------------------------------------------------------

BoundaryRegistry = dict[str, StandardBoundaryRule]
"""Mutable mapping from ``"from_domain->to_domain"`` keys to rules."""


def build_boundary_key(from_domain: str, to_domain: str) -> str:
    """Build the registry lookup key for a from/to domain pair."""
    return f"{from_domain}->{to_domain}"


def build_boundary_registry(
    rules: list[StandardBoundaryRule] | tuple[StandardBoundaryRule, ...],
) -> BoundaryRegistry:
    """
    Build a :data:`BoundaryRegistry` from a list of rules.
    For ``bidirectional`` rules, entries are created in both directions.

    :param rules: Boundary rules to index.
    :returns: Dict keyed by ``"from->to"`` strings.
    :raises ValueError: If two rules share the same directional key.
    """
    registry: BoundaryRegistry = {}
    for rule in rules:
        forward_key = build_boundary_key(rule.from_domain, rule.to_domain)
        if forward_key in registry:
            raise ValueError(
                f"Boundary rule conflict: a rule for '{forward_key}' is already registered. "
                f"Existing rule: '{registry[forward_key].name}', new rule: '{rule.name}'."
            )
        registry[forward_key] = rule

        if rule.direction == "bidirectional":
            reverse_key = build_boundary_key(rule.to_domain, rule.from_domain)
            if reverse_key in registry:
                raise ValueError(
                    f"Boundary rule conflict: bidirectional rule '{rule.name}' would overwrite "
                    f"existing rule for '{reverse_key}'."
                )
            registry[reverse_key] = rule

    return registry

File: typescript/src/boundary.ts
// SPDX-License-Identifier: BSL-1.1
// Copyright (c) 2026 MuVeraAI Corporation

/**
 * @module boundary
 * BoundaryRule definitions for context-firewall.
 *
 * A BoundaryRule specifies which data types may or may not cross between
 * two domains. Rules are evaluated by DataInspector before any crossing
 * is permitted.
 */

import { z } from "zod";
import {
  BoundaryDirection,
  BoundaryDirectionSchema,
  DataClassification,
} from "./types.js";

// ---------------------------------------------------------------------------
// BoundaryRule interface
// ---------------------------------------------------------------------------

/**
 * Defines the crossing rules between two domains.
 *
 * @example
 * ```typescript
 * const healthToWork: BoundaryRule = {
 *   name: "health->work",
 *   fromDomain: "health",
 *   toDomain: "work",
 *   direction: "one-way",
 *   allowedDataTypes: [],
 *   blockedDataTypes: ["medical", "prescription", "diagnosis", "mental-health"],
 *   evaluate(classification) {
 *     return classification.detectedTypes.every(
 *       (t) => !this.blockedDataTypes.includes(t)
 *     );
 *   },
 * };
 * ```
 */
export interface BoundaryRule {
  /** Unique identifier for this rule, used in `FirewallDecision.appliedRuleName`. */
  readonly name: string;
  /** The originating domain. */
  readonly fromDomain: string;
  /** The destination domain. */
  readonly toDomain: string;
  /**
   * Direction of the rule.
   * - `one-way`: enforced only from `fromDomain` to `toDomain`.
   * - `bidirectional`: enforced in both directions.
   */
  readonly direction: BoundaryDirection;
  /**
   * Data types explicitly allowed to cross.
   * An empty array means "no explicit allowlist — use blockedDataTypes only".
   */
  readonly allowedDataTypes: ReadonlyArray<string>;
  /**
   * Data types that are never permitted to cross, regardless of other rules.
   */
  readonly blockedDataTypes: ReadonlyArray<string>;
  /**
   * Evaluate whether a given classification is permitted to cross.
   *
   * @param classification - The classification of the data being inspected
   * @returns `true` if crossing is allowed, `false` if it must be blocked
   */
  evaluate(classification: DataClassification): boolean;
}

// ---------------------------------------------------------------------------
// Zod schema for the serialisable portion of a BoundaryRule
// ---------------------------------------------------------------------------

export const BoundaryRuleConfigSchema = z.object({
  name: z.string().min(1),
  fromDomain: z.string().min(1),
  toDomain: z.string().min(1),
  direction: BoundaryDirectionSchema,
  allowedDataTypes: z.array(z.string()),
  blockedDataTypes: z.array(z.string()),
});

export type BoundaryRuleConfig = z.infer<typeof BoundaryRuleConfigSchema>;

// ---------------------------------------------------------------------------
// Factory: create a standard BoundaryRule from config
// ---------------------------------------------------------------------------

/**
 * Build a `BoundaryRule` from a plain config object.
 * The `evaluate` method enforces the following logic:
 *
 * 1. If `blockedDataTypes` is non-empty, any detected type in that list blocks the crossing.
 * 2. If `allowedDataTypes` is non-empty, only types in that list are permitted; any
 *    detected type not on the allowlist blocks the crossing.
 * 3. If both lists are empty, the crossing is allowed (open rule).
 *
 * @param config - Serialisable rule configuration
 * @returns A fully formed `BoundaryRule`
 */
export function createBoundaryRule(config: BoundaryRuleConfig): BoundaryRule {
  const validated = BoundaryRuleConfigSchema.parse(config);

  return {
    name: validated.name,
    fromDomain: validated.fromDomain,
    toDomain: validated.toDomain,
    direction: validated.direction,
    allowedDataTypes: validated.allowedDataTypes,
    blockedDataTypes: validated.blockedDataTypes,

    evaluate(classification: DataClassification): boolean {
      for (const detectedType of classification.detectedTypes) {
        // Blocked list takes priority
        if (this.blockedDataTypes.includes(detectedType)) {
          return false;
        }
        // Allowlist enforcement: if an allowlist exists, type must be on it
        if (
          this.allowedDataTypes.length > 0 &&
          !this.allowedDataTypes.includes(detectedType)
        ) {
          return false;
        }
      }
      return true;
    },
  };
}

// ---------------------------------------------------------------------------
// Default built-in boundary rules
// ---------------------------------------------------------------------------

/**
 * health -> work: no medical data crosses into professional context.
 */
export const HEALTH_TO_WORK_RULE: BoundaryRule = createBoundaryRule({
  name: "health->work",
  fromDomain: "health",
  toDomain: "work",
  direction: "one-way",
  allowedDataTypes: [],
  blockedDataTypes: [
    "medical",
    "prescription",
    "diagnosis",
    "mental-health",
    "clinical",
    "laboratory",
  ],
});

/**
 * financial -> work: no financial account or tax data crosses into work context.
 */
export const FINANCIAL_TO_WORK_RULE: BoundaryRule = createBoundaryRule({
  name: "financial->work",
  fromDomain: "financial",
  toDomain: "work",
  direction: "one-way",
  allowedDataTypes: [],
  blockedDataTypes: [
    "banking",
    "credit-card",
    "tax",
    "investment",
    "insurance-financial",
    "salary",
  ],
});

/**
 * health -> personal: medical data may not cross into personal domain by default.
 * Rationale: keeps HIPAA-regulated data isolated even within the user's own contexts.
 */
export const HEALTH_TO_PERSONAL_RULE: BoundaryRule = createBoundaryRule({
  name: "health->personal",
  fromDomain: "health",
  toDomain: "personal",
  direction: "one-way",
  allowedDataTypes: [],
  blockedDataTypes: [
    "medical",
    "prescription",
    "diagnosis",
    "mental-health",
    "clinical",
    "laboratory",
  ],
});

/**
 * financial -> personal: core financial account data may not flow into personal context.
 */
export const FINANCIAL_TO_PERSONAL_RULE: BoundaryRule = createBoundaryRule({
  name: "financial->personal",
  fromDomain: "financial",
  toDomain: "personal",
  direction: "one-way",
  allowedDataTypes: [],
  blockedDataTypes: [
    "banking",
    "credit-card",
    "tax",
    "investment",
    "salary",
  ],
});

/**
 * personal -> work: personal and family data must not enter the work domain.
 */
export const PERSONAL_TO_WORK_RULE: BoundaryRule = createBoundaryRule({
  name: "personal->work",
  fromDomain: "personal",
  toDomain: "work",
  direction: "one-way",
  allowedDataTypes: [],
  blockedDataTypes: [
    "family",
    "relationship",
    "home-address",
    "personal-contact",
  ],
});

/**
 * The complete set of default built-in boundary rules.
 * Registered automatically by `ContextFirewall` unless `skipDefaultBoundaries: true`.
 */
export const DEFAULT_BOUNDARY_RULES: ReadonlyArray<BoundaryRule> = [
  HEALTH_TO_WORK_RULE,
  FINANCIAL_TO_WORK_RULE,
  HEALTH_TO_PERSONAL_RULE,
  FINANCIAL_TO_PERSONAL_RULE,
  PERSONAL_TO_WORK_RULE,
];

// ---------------------------------------------------------------------------
// Boundary registry
// ---------------------------------------------------------------------------

/**
 * A map of boundary rules keyed by `"fromDomain->toDomain"` lookup strings.
 * Both `one-way` and `bidirectional` rules are indexed in both directions
 * when applicable.
 */
export type BoundaryRegistry = ReadonlyMap<string, BoundaryRule>;

/**
 * Build a registry lookup key for a from/to domain pair.
 */
export function buildBoundaryKey(fromDomain: string, toDomain: string): string {
  return `${fromDomain}->${toDomain}`;
}

/**
 * Build a `BoundaryRegistry` from an array of `BoundaryRule` objects.
 * For `bidirectional` rules, entries are created in both directions.
 *
 * @param rules - Boundary rules to index
 * @returns Immutable Map keyed by `"from->to"` strings
 * @throws {Error} If two rules share the same directional key
 */
export function buildBoundaryRegistry(
  rules: ReadonlyArray<BoundaryRule>
): BoundaryRegistry {
  const registry = new Map<string, BoundaryRule>();

  for (const rule of rules) {
    const forwardKey = buildBoundaryKey(rule.fromDomain, rule.toDomain);

    if (registry.has(forwardKey)) {
      throw new Error(
        `Boundary rule conflict: a rule for '${forwardKey}' is already registered. ` +
          `Rule names: existing='${registry.get(forwardKey)!.name}', new='${rule.name}'.`
      );
    }
    registry.set(forwardKey, rule);

    if (rule.direction === "bidirectional") {
      const reverseKey = buildBoundaryKey(rule.toDomain, rule.fromDomain);
      if (registry.has(reverseKey)) {
        throw new Error(
          `Boundary rule conflict: bidirectional rule '${rule.name}' would overwrite ` +
            `existing rule for '${reverseKey}'.`
        );
      }
      registry.set(reverseKey, rule);
    }
  }

  return registry;
}
